### ISSUE TITLE
tls-related segfault in jemalloc 5.1.0 after upgrade from 4.1.0, redhat 6 with gcc 6.3

### ISSUE THREAD

**Issue by dposton80** [jemalloc/jemalloc#1234]
We have an occasional crash in jemalloc that usually occurs with the following callstack:

```
je_arena_stats_merge at src/arena.c:312
ctl_arena_stats_amerge at src/ctl.c:694
ctl_arena_refresh at src/ctl.c:860
ctl_refresh at src/ctl.c:1484
je_ctl_byname at src/ctl.c:1173
je_mallctl at src/jemalloc.c:2913
```

Our application calls _mallctl( "stats.allocated" ) quite regularly. However I have seen it crash in je_malloc as well (don't have the exact callstack I'm afraid). It looks like thread local storage is corrupted in some way.

About one in 50 times, instead of crashing, it goes into an infinite loop in je_arena_stats_merge. Using gdb it seems to be stuck looping over thread caches, gathering stats.

It takes the application between 1 and 20 minutes to crash. It's a multithreaded program that is stressing malloc reasonably hard.

The application is using je_malloc as a shared object built with --with-jemalloc-prefix=je_. We then have a set of trampoline functions that define malloc, free, etc and redirect to je_malloc, je_free after adjusting some thread-local memory stats. These functions use pthread_getspecific to get a pointer to a structure, and then increment counters.

Building jemalloc with --enable-debug gives an assertion seemingly after the segfault has occurred (so may be a red herring). The output looks like:
```
*** Segmentation fault
Register dump:
RAX: ...
<jemalloc>: Should own 0 locks of rank >= 1: ctl(1) tcache_ql(12)
```
Some configurations of the application, using the same binaries, would not crash at all. The configurations that crashed would do so consistently. It does not crash with jemalloc 4.1.0.

Removing the trampoline functions, and running with 'regular' jemalloc (no prefix), stops the crash. The --enable-debug build of jemalloc does not report any assertions in this case. asan did not report any problems. asan was run with the same workload as would normally crash.

Turning off thread cache (JE_MALLOC_CONF=tcache:false) also stopped it crashing.

I am suspicious that the application is doing something wrong, but its only interaction with tls is calling pthread_[get/set]specific, and beyond that everything is on the heap (and thus verified to be ok with asan). I am also suspicious of the trampoline functions - but why would they be fine with tcache:false?

So I suspect this is an issue with jemalloc 5 - any tips on debugging this would be appreciated.

**Comment by interwq** [comment]
Can you please share the stack trace with --enable-debug? It may be a jemalloc bug based on the lock ranking assertion.

**Comment by dposton80** [comment]
It looks like the assert is caused by recursive malloc from our segfault handler, so it's probably not relevant (but I guess it proves that whatever the cause of this, there's no jemalloc assert for it):
```
#0  0x00007f092f12293d in je_malloc_printf ()
   from ./jemalloc-5.1.0/libjemalloc.so.2
#1  0x00007f092f148611 in witness_depth_error_impl ()
   from ./jemalloc-5.1.0/libjemalloc.so.2
#2  0x00007f092f0c3004 in witness_assert_depth_to_rank ()
    at include/jemalloc/internal/witness.h:259
#3  0x00007f092f0c302f in witness_assert_depth ()
    at include/jemalloc/internal/witness.h:265
#4  0x00007f092f0c304f in witness_assert_lockless ()
    at include/jemalloc/internal/witness.h:270
#5  0x00007f092f0cb598 in check_entry_exit_locking () at src/jemalloc.c:686
#6  0x00007f092f0cebf6 in imalloc_body ()
   from ./jemalloc-5.1.0/libjemalloc.so.2
#7  0x00007f092f0cf1b7 in imalloc ()
   from ./jemalloc-5.1.0/libjemalloc.so.2
#8  0x00007f092f0cf2c8 in je_malloc ()
   from ./jemalloc-5.1.0/libjemalloc.so.2
#9  0x00000000004165bf in malloc ()
    at .src/tkdbs/src/MemoryHooks.cpp:21
#10 0x000000366ae0cb9d in _dl_map_object_deps ()
   from /lib64/ld-linux-x86-64.so.2
#11 0x000000366ae12a91 in dl_open_worker () from /lib64/ld-linux-x86-64.so.2
#12 0x000000366ae0e1a6 in _dl_catch_error () from /lib64/ld-linux-x86-64.so.2
---Type <return> to continue, or q <return> to quit---
#13 0x000000366ae124ea in _dl_open () from /lib64/ld-linux-x86-64.so.2
#14 0x000000366b726300 in do_dlopen () from /lib64/libc.so.6
#15 0x000000366ae0e1a6 in _dl_catch_error () from /lib64/ld-linux-x86-64.so.2
#16 0x000000366b726457 in __libc_dlopen_mode () from /lib64/libc.so.6
#17 0x000000366b6fe855 in init () from /lib64/libc.so.6
#18 0x000000366be0cb23 in pthread_once () from /lib64/libpthread.so.0
#19 0x000000366b6fe954 in backtrace () from /lib64/libc.so.6
#20 0x00007f0932419f93 in catch_segfault () from /lib64/libSegFault.so
#21 <signal handler called>
#22 0x00007f092f0dc0e6 in je_arena_stats_merge ()
   from ./jemalloc-5.1.0/libjemalloc.so.2
#23 0x00007f092f0f18d4 in ctl_arena_stats_amerge ()
   from ./jemalloc-5.1.0/libjemalloc.so.2
#24 0x00007f092f0f27f0 in ctl_arena_refresh ()
   from ./jemalloc-5.1.0/libjemalloc.so.2
#25 0x00007f092f0f2b15 in ctl_refresh ()
   from ./jemalloc-5.1.0/libjemalloc.so.2
#26 0x00007f092f0f3891 in epoch_ctl ()
   from ./jemalloc-5.1.0/libjemalloc.so.2
#27 0x00007f092f0f3406 in je_ctl_byname ()
   from ./jemalloc-5.1.0/libjemalloc.so.2
#28 0x00007f092f0d357e in je_mallctl ()
   from ./jemalloc-5.1.0/libjemalloc.so.2
```

The crash is in the ql_foreach block in arena_stats_merge (which is also where the infinite loop happens, when it does)

**Comment by interwq** [comment]
@dposton80 : yep the crash points to heap corruption by application. I'd suggest trying sanitizers like ASAN to investigate further. Closing for now. Let us know if anything else.

**Comment by interwq** [comment]
As you mentioned `Removing the trampoline functions, and running with 'regular' jemalloc (no prefix), stops the crash. ` I'd be careful to check if any mixed allocator activity, since that could easily trigger the behaviors mentioned here.

**Comment by dposton80** [comment]
Thanks for looking into it. Will follow up if I find anything. Did verify (via dumping allocs to file) that there was no mixed malloc activity, but couldn't run it for long in this mode.

**Comment by dposton80** [comment]
I think I found what this was. It turned out to be not quite mixed alloc activity, but calling into jemalloc before the .so had been fully initialized (I guess whilst other .sos are loading, but before ldd had got round to loading jemalloc). It worked fine with thread caches turned off, but recently we hit another crash when we introduced C++ 11 threads to the application, which affects tls in some way. This got fixed by adding jemalloc to LD_PRELOAD. Haven't tried turning thread caches back on yet though!
